vita: multiple control enhancements (see below)

rsn8887 opened the pull request:
- Bluetooth keyboard and mouse support
- On-screen keyboard by pressing start, useful to enter player name etc.
- Analog joystick mouse control via left stick
- Touchpad-style mouse control with gestures (short tap for left click, etc.)
- Map right analog stick to cursor keys
- Map L/R to mousebuttons, and square/triangle to page up/down
- Press select to toggle between direct touch (pointer jumps to finger, nothing else) or indirect touch (drag pointer and single/touble tap to left/right click like on a touchpad)

rsn8887 committed:
vita: multiple control enhancements (see below)

- Bluetooth keyboard and mouse support
- On-screen keyboard by pressing start, useful to enter player name etc.
- Analog joystick mouse control via left stick
- Touchpad-style mouse control with gestures (short tap for left click, etc.)
- Map right analog stick to cursor keys
- Map L/R to mousebuttons, and square/triangle to page up/down

rsn8887 commented:
I had to create a dummy renderer and window because SDL only generates events when video subsystem is running.

This is only to get events working, the renderer is never used for rendering. The optimized vita2d routines are still used for all the display stuff.

rsn8887 commented:
For some reason my bluetooth mouse requires one left mouse click to activate it, and then it works just fine with this.

devnoname120 commented:
How much work would it be to use `SceCtrl` instead of the SDL?
I would like to rely less on the SDL, not more.

rsn8887 commented:
To get all these features? A lot, and IMO pointless because it is already implemented in SDL and causes no overhead or slowdown anyways. 

For example, physical keyboard and mouse polling which is already implemented in SDL (and in the game). No need to duplicate and modify that code.

I agree that the video stuff in SDL can be slow and it is often good to use vita2d directly, but there’s no point on re-inventing the wheel for input and events, which is already fast and convenient as it is.

devnoname120 reviewed (changes requested):
I'm not fundamentally opposed to using the SDL for input because it could be reused in other ports. However the code as is has several coding-style issues:
- Sometimes it uses the name `psp2`, and sometimes `vita`. Please be consistent and use `vita` everywhere (file names, function names, etc.).
- The coding style is not respected (it uses tabs instead of spaces for example).
- There is switch-related code even though this branch is dedicated to the vita port, and in addition it's located in `vita` platform files.
- `#ifdef __cplusplus`, the whole code is C so this is unnecessary.
- Some parts of the code have 8(!) levels of indentation. This makes the code tremendously hard to follow.
- Some conditionals are superfluous ([this](https://github.com/devnoname120/julius/pull/1/files#diff-7ee2ae5edb44cc0d54c530b6a6430691R231) condition and [this](https://github.com/devnoname120/julius/pull/1/files#diff-7ee2ae5edb44cc0d54c530b6a6430691R266) are the same). 
- Hardcoded constants throughout the code (`960`, `544`, etc.)
- Two-letter variables.
- Etc.

Since my ambition is to do a pull request on the main repo, I cannot merge this in its current state.

rsn8887 commented:
Thanks for the comments.

I think I fixed all the formatting issues, apart from the logic of the touch code.

Considering the logic of the touch code, well, it is complicated to do touch gestures correctly including multiple fingers etc. I learned that. It took me a while to get it right and it is not trivial. So I think the reason it is hard to read is just because it is complicated. It is easy to do crappy / half-functioning touch controls. But it is hard to do them right. With crappy/half-functioning touch controls, it becomes very frustrating to do simple things like left click/right click/drag with your fingers without sometimes erroneously generating clicks while moving the pointer, etc.

Considering the superfluous conditionals you mentioned. I don't see those as superfluous. Those are there to make it easier to read. One conditional is there to check whether a multifinger drag is occuring. The other conditional is there to check what mouse motion should be generated. These are two different meaningful states that happen to correspond to similar conditions.

By using separate conditionals and code blocks for those two things, the code becomes easier, not harder, to follow.

I don't see the point of lumping the conditionals together and combining two quite different things. It makes it even harder to read and maintain in the future if any change is requested that might only affect one of these. And saving a single if instruction per frame is not worth that in my opinion.

I hope the touch code can be left as is, because it took me long enough to get it right. A refactor would require a lot of work debugging it again, etc.

I agree the code is not pretty, but it is functional. Unless there's some code policy in the main project limiting if-clause pyramids I hope it might be acceptable.

rsn8887 commented:
I fixed the remaining camelCase variables...

rsn8887 commented:
changed display_width/height to defined DISPLAY_WIDTH/HEIGHT in the Vita code.

rsn8887 commented:
Some more whitespace fixes and function naming in vita_keyboard.c is now also fixed.

I think this is about everything I could find.

rsn8887 commented:
Found a few more whitespaces I just fixed them.

rsn8887 commented:
renamed last_mx/my to last_mouse_x/y for clarity.

rsn8887 commented:
With these changes, I think it is now much more style-conforming and readable. Please take a look @devnoname120 whenever you find a few minutes, and see if it is acceptable now.

devnoname120 commented on src/platform/vita/vita_touch.c:
```diff
@@ -0,0 +1,343 @@
+    // pointer motion: single finger drag
+    // left button drag and drop: dual finger drag
+    // right button drag and drop: triple finger drag
+    if (event->type == SDL_FINGERDOWN || event->type == SDL_FINGERUP || event->type == SDL_FINGERMOTION) {
```
```suggestion
    if (event->type != SDL_FINGERDOWN && event->type != SDL_FINGERUP && event->type != SDL_FINGERMOTION)
        return;
```

Please invert this way all conditionals that end at the end of the function's body.

rsn8887 replied:
Fixed in 35f42ab00e2b28a49e966a752eb3016a334d53bb

devnoname120 commented on src/platform/vita/vita_touch.c:
```diff
@@ -0,0 +1,343 @@
+    if (event->type == SDL_FINGERDOWN || event->type == SDL_FINGERUP || event->type == SDL_FINGERMOTION) {
+        // front (0) or back (1) panel
+        SDL_TouchID port = event->tfinger.touchId;
+        if (port < SCE_TOUCH_PORT_MAX_NUM && port >= 0) {
```
This condition is redundant.

rsn8887 replied:
Fixed in 35f42ab00e2b28a49e966a752eb3016a334d53bb

devnoname120 commented on src/platform/vita/vita_touch.c:
```diff
@@ -0,0 +1,343 @@
+    // make sure each finger is not reported down multiple times
+    for (int i = 0; i < MAX_NUM_FINGERS; i++) {
+        if (finger[port][i].id == id) {
+            finger[port][i].id = -1;
```
Use a constant for `-1` to make it clearer what is represents.

rsn8887 replied:
Fixed in 35f42ab00e2b28a49e966a752eb3016a334d53bb

devnoname120 commented on src/platform/vita/vita_input.c:
```diff
@@ -0,0 +1,428 @@
+
+static int hires_dx = 0;
+static int hires_dy = 0;
+static int left_pressed;
```
Please explicitly set those fields to `0`.

rsn8887 replied:
Fixed in 35f42ab00e2b28a49e966a752eb3016a334d53bb

devnoname120 commented on src/platform/vita/vita_input.c:
```diff
@@ -0,0 +1,428 @@
+    int left_x = SDL_JoystickGetAxis(joy, 0);
+    int left_y = SDL_JoystickGetAxis(joy, 1);
+    rescale_analog(&left_x, &left_y, 3000);
+    hires_dx += left_x;
```
What does `hires_dx` mean?

rsn8887 replied:
Fixed in 35f42ab00e2b28a49e966a752eb3016a334d53bb

devnoname120 commented on src/platform/vita/vita_touch.c:
```diff
@@ -0,0 +1,343 @@
+{
+    for (int port = 0; port < SCE_TOUCH_PORT_MAX_NUM; port++) {
+        for (int i = 0; i < 2; i++) {
+            if (simulated_click_start_time[port][i] != 0) {
```
```suggestion
            if (simulated_click_start_time[port][i] == 0)
                continue;
```

You get the idea.

rsn8887 replied:
Fixed in 35f42ab00e2b28a49e966a752eb3016a334d53bb

devnoname120 commented on src/platform/vita/vita_touch.c:
```diff
@@ -0,0 +1,343 @@
+                if (finger[port][i].id == id) {
+                    for (int j = 0; j < MAX_NUM_FINGERS; j++) {
+                        if (finger[port][j].id >= 0 && (i != j) ) {
+                            if (finger[port][j].time_last_down < finger[port][i].time_last_down) {
```
Reduce nesting here as well.

rsn8887 replied:
Fixed in 35f42ab00e2b28a49e966a752eb3016a334d53bb

devnoname120 commented on src/platform/vita/vita_input.h:
```diff
@@ -0,0 +1,24 @@
+void vita_handle_virtual_keyboard(void);
+void vita_handle_repeat_keys(void);
+
+#define PAD_UP 8
```
```suggestion
#define VITA_PAD_UP 8
```

devnoname120 replied:
Also, I suggest using an `enum` since those values make sense together.

rsn8887 replied:
Fixed in 35f42ab00e2b28a49e966a752eb3016a334d53bb

devnoname120 commented on src/platform/vita/vita_keyboard.c:
```diff
@@ -0,0 +1,173 @@
+static int init_ime_dialog(char *title, const char *initial_text, int max_text_length, int type, int option)
+{
+    if (ime_dialog_running)
+    return -1;
```
Indentation.

rsn8887 replied:
Fixed in 35f42ab00e2b28a49e966a752eb3016a334d53bb

devnoname120 commented on src/platform/vita/vita_keyboard.c:
```diff
@@ -0,0 +1,173 @@
+
+static int update_ime_dialog(void) {
+    if (!ime_dialog_running)
+    return IME_DIALOG_RESULT_NONE;
```
Here as well.

rsn8887 replied:
Fixed in 35f42ab00e2b28a49e966a752eb3016a334d53bb

devnoname120 commented on src/platform/vita/vita_keyboard.c:
```diff
@@ -0,0 +1,173 @@
+        ime_init_apputils = 1;
+    }
+    if (multiline)
+    init_ime_dialog(title, initial_text, maxLen, SCE_IME_TYPE_BASIC_LATIN, SCE_IME_OPTION_MULTILINE);
```
Indentation.

rsn8887 replied:
Fixed in 35f42ab00e2b28a49e966a752eb3016a334d53bb

devnoname120 commented on src/platform/vita/vita_keyboard.c:
```diff
@@ -0,0 +1,173 @@
+    else
+    init_ime_dialog(title, initial_text, maxLen, SCE_IME_TYPE_BASIC_LATIN, 0);
+
+    bool done;
```
```suggestion
    bool done = false;
```

rsn8887 replied:
Fixed in 35f42ab00e2b28a49e966a752eb3016a334d53bb

devnoname120 commented on src/platform/vita/vita_keyboard.c:
```diff
@@ -0,0 +1,173 @@
+    init_ime_dialog(title, initial_text, maxLen, SCE_IME_TYPE_BASIC_LATIN, 0);
+
+    bool done;
+    do {
```
```suggestion
    while (!done) {
```

rsn8887 replied:
Fixed in 35f42ab00e2b28a49e966a752eb3016a334d53bb

devnoname120 commented on src/platform/vita/vita_keyboard.c:
```diff
@@ -0,0 +1,173 @@
+
+        int ime_result = update_ime_dialog();
+        if (ime_result == IME_DIALOG_RESULT_FINISHED) {
+            name = (char *)get_ime_dialog_input_text_utf8();
```
Doesn't it work to return `name` here instead of using the `done` boolean?

rsn8887 replied:
No, I remember testing this, and the vita2d_commands have to be executed, otherwise I got gfx flicker in Rocks'n'Diamonds at least.

devnoname120 commented on src/platform/vita/vita_touch.c:
```diff
@@ -0,0 +1,343 @@
+
+static Touch finger[SCE_TOUCH_PORT_MAX_NUM][MAX_NUM_FINGERS]; // keep track of finger status
+
+typedef enum DraggingType {
```
```suggestion
typedef enum {
```

rsn8887 replied:
Fixed in 35f42ab00e2b28a49e966a752eb3016a334d53bb

devnoname120 commented on src/platform/vita/vita_touch.c:
```diff
@@ -0,0 +1,343 @@
+
+typedef int bool;
+#define false 0;
+#define true 1;
```
Really? Why not use `stdbool.h`?

rsn8887 replied:
Fixed in 35f42ab00e2b28a49e966a752eb3016a334d53bb

devnoname120 commented on src/platform/vita/vita_input.c:
```diff
@@ -0,0 +1,428 @@
+            SDL_PushEvent(&up_event);
+        }
+        int i=0;
+        while (text[i]!=0 && i<599) {
```
```suggestion
        for (int i=0; i < 599 && text[i] != 0; i++) {
```

Or maybe
```suggestion
        for (int i=0; i < 599; i++) {
            if (text[i] == 0)
                break;
```

rsn8887 replied:
Fixed in 35f42ab00e2b28a49e966a752eb3016a334d53bb

devnoname120 commented on src/platform/vita/vita_input.c:
```diff
@@ -0,0 +1,428 @@
+                            break;
+                    }
+                }
+                break;
```
Lots of redundant code in the switch statement. Could you refactor this?

You could for instance have an array containing pressed keys and indexed by `PAD_X` instead of `right_pressed`, `left_pressed`, etc. Then you can have a function taking `PAD_X` as input and doing the 5 statements. You can have another array to do the mapping between `PAD_X` and `SDLK_X`.

rsn8887 replied:
Fixed in 35f42ab00e2b28a49e966a752eb3016a334d53bb

devnoname120 commented on src/platform/vita/vita_input.c:
```diff
@@ -0,0 +1,428 @@
+            down_event.type = SDL_KEYDOWN;
+            down_event.key.keysym.sym = SDLK_BACKSPACE;
+            down_event.key.keysym.mod = 0;
+            SDL_PushEvent(&down_event);
```
Put this in a helper function.

rsn8887 replied:
Fixed in 35f42ab00e2b28a49e966a752eb3016a334d53bb

devnoname120 commented on src/platform/vita/vita_input.c:
```diff
@@ -0,0 +1,428 @@
+#include <math.h>
+
+#include "vita_keyboard.h"
+#define DISPLAY_WIDTH 960
```
Defined several times, put it in a header file and prefix it with `VITA`.

rsn8887 replied:
Fixed in 35f42ab00e2b28a49e966a752eb3016a334d53bb

devnoname120 reviewed (commented):
It's better, but there is still a good amount of refactoring needed.

------------------

> combining two quite different things

Usually when you have two 'quite different things' next to each other, it's a good hint that you need extra functions.

Besides, I understand your point that it was hard to get it working so you'd rather not touch it. However if down the road something needs to be fixed, it will be impossible to understand what's going on. Actually there are so many moving parts that I'd very surprised if there aren't any weird bugs creeping inside.

rsn8887 committed:
vita: improve code formatting and readability

rsn8887 commented:
I think I fixed everything now.... with commit  35f42ab00e2b28a49e966a752eb3016a334d53bb

rsn8887 commented:
I can happily report that everything still works great (touch control, analog stick controls click gestures etc.), even after all the re-factoring.

I'd it should ready or close to ready to be merged, @devnoname120 if you want to take another look and or merge.

devnoname120 commented:
Awesome, I still have a few comments to do but we're close.

devnoname120 commented on src/platform/vita/vita_input.c:
```diff
@@ -0,0 +1,375 @@
+                            break;
+                        case VITA_PAD_CROSS:
+                        case VITA_PAD_R: // intentional fallthrough
+                            event->type = SDL_MOUSEBUTTONDOWN;
```
I suggest moving this into a `vita_button_to_sdlmouse_event` function.

rsn8887 replied:
Ok, done in https://github.com/rsn8887/julius/commit/23eb3377768198b00239bcca3585ffa3e2128063

devnoname120 commented on src/platform/vita/vita_input.c:
```diff
@@ -0,0 +1,375 @@
+                last_mouse_y = event->motion.y;
+                break;
+            case SDL_JOYBUTTONDOWN:
+                if (event->jbutton.which==0) { // Only Joystick 0 controls the game
```
```suggestion
                if (event->jbutton.which != 0) // Only Joystick 0 controls the game
                    break;
```

rsn8887 replied:
Ok, done in https://github.com/rsn8887/julius/commit/23eb3377768198b00239bcca3585ffa3e2128063

devnoname120 commented on src/platform/vita/vita_input.c:
```diff
@@ -0,0 +1,375 @@
+                }
+                break;
+            case SDL_JOYBUTTONUP:
+                if (event->jbutton.which==0) {// Only Joystick 0 controls the game
```
```suggestion
                if (event->jbutton.which != 0) // Only Joystick 0 controls the game
                    break;
```

rsn8887 replied:
Ok, done in https://github.com/rsn8887/julius/commit/23eb3377768198b00239bcca3585ffa3e2128063

devnoname120 commented on src/platform/vita/vita_input.c:
```diff
@@ -0,0 +1,375 @@
+                            break;
+                        case VITA_PAD_CROSS:
+                        case VITA_PAD_R: // intentional fallthrough
+                            event->type = SDL_MOUSEBUTTONUP;
```
Move to `vita_button_to_sdlmouse_event`.

rsn8887 replied:
Ok, done in https://github.com/rsn8887/julius/commit/23eb3377768198b00239bcca3585ffa3e2128063

devnoname120 commented on src/platform/vita/vita_input.c:
```diff
@@ -0,0 +1,375 @@
+void vita_start_text_input(char *initial_text, int multiline)
+{
+    char *text = vita_keyboard_get("Enter New Text:", initial_text, 600, multiline);
+    if (text != NULL)  {
```
```suggestion
    if (text == NULL) 
        return;
```

rsn8887 replied:
Ok, done in https://github.com/rsn8887/julius/commit/23eb3377768198b00239bcca3585ffa3e2128063

devnoname120 commented on src/platform/vita/vita_touch.c:
```diff
@@ -0,0 +1,332 @@
+
+#include "vita_touch.h"
+
+#include "math.h"
```
```suggestion
#include <math.h>
```

rsn8887 replied:
Ok, done in https://github.com/rsn8887/julius/commit/23eb3377768198b00239bcca3585ffa3e2128063

devnoname120 commented on src/platform/vita/vita_touch.c:
```diff
@@ -0,0 +1,332 @@
+#include <psp2/kernel/processmgr.h>
+#include "vita.h"
+
+#define NO_TOUCH -1 // finger id setting if finger is not touching the screen
```
Move this below the headers.

rsn8887 replied:
Ok, done in https://github.com/rsn8887/julius/commit/23eb3377768198b00239bcca3585ffa3e2128063

devnoname120 commented on src/platform/vita/vita_touch.c:
```diff
@@ -0,0 +1,332 @@
+#ifdef __vita__
+    // prevent suspend
+    sceKernelPowerTick(SCE_KERNEL_POWER_TICK_DISABLE_AUTO_SUSPEND);
+    sceKernelPowerTick(SCE_KERNEL_POWER_TICK_DISABLE_OLED_OFF);
```
Why preventing screen off?

rsn8887 replied:
Leftover from ScummVM Vita, removed in https://github.com/rsn8887/julius/commit/23eb3377768198b00239bcca3585ffa3e2128063

rsn8887 replied:
Oops I removed both. Fixed in 1a45a77314ab98488d51a58aaf8e7eda55a22567

rsn8887 commented:
@devnoname120 did you get a chance to look over it again yet? Are you maybe waiting to for the official touch support to be merged? I can easily remove the touch support from here if that is what you like. It is just removing one file and one line of code.

rsn8887 committed:
vita: more code formatting and readability improvements

rsn8887 committed:
vita: prevent auto_suspend

rsn8887 committed:
vita: direct touch mode (toggle via select), if clause formatting

rsn8887 committed:
vita: fix CMAKELISTS.txt env var error

rsn8887 commented:
@devnoname120 : I think this is ready or at least very close now. Of course, it should be rebased on merge. I left the commit noise in there for now to aid review.  
- I fixed all the issues from the last review.
- I added missing if clause brackets.
- I re-added the old touch mode (pointer jumps to finger). It can be toggled via SELECT
- I fixed that CMAKELISTS.TXT env var you mentioned on discord.

Please tell me if there's anything else I can do to get this merged :)

Note: I kept the touchpad touch mode as default. Feel free to change it in line 27 of vita_input.c:

`int direct_touch_mode = 0; // 0: drag the pointer and click like on a touchpad (default), 1: pointer jumps to finger`

devnoname120 commented:
Upstream pull request: https://github.com/bvschaik/julius/pull/114.
